src/Features/Core/Portable/Navigation/INavigableItem.cs

```diff
@@ -45,5 +48,39 @@ internal interface INavigableItem
         bool IsStale { get; }
 
         ImmutableArray<INavigableItem> ChildItems { get; }
+
+        public record NavigableDocument(NavigableProject Project, string Name, string? FilePath, IReadOnlyList<string> Folders, DocumentId Id)
+        {
+            public required Workspace Workspace { get; init; }
+
+            public static NavigableDocument FromDocument(Document document)
+                => new(NavigableProject.FromProject(document.Project), document.Name, document.FilePath, document.Folders, document.Id) { Workspace = document.Project.Solution.Workspace };
+
+            internal async ValueTask<Document> GetDocumentAsync(Solution solution, CancellationToken cancellationToken)
+            {
+                if (solution.GetDocument(Id) is { } document)
+                    return document;
+
+                return await solution.GetRequiredDocumentAsync(Id, includeSourceGenerated: true, cancellationToken).ConfigureAwait(false);
```

Comment:
📝 I'm not sure this is the correct approach. Ideally, we would retain information from the input document about whether it's a source generated document, and use that information here to only specify `includeSourceGenerated` if necessary.